packages/expo-updates/ios/EXUpdates/CodeSigning/EXUpdatesCertificateChain.swift

```diff
@@ -0,0 +1,194 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+import Foundation
+import ASN1Decoder
+
+typealias Certificate = (SecCertificate, X509Certificate)
+
+/**
+ * Full certificate chain for verifying code signing.
+ * The chain should look like the following:
+ *    0: code signing certificate
+ *    1...n-1: intermediate certificates
+ *    n: root certificate
+ *
+ * Requirements:
+ * - Length(certificateChain) > 0
+ * - certificate chain is valid and each certificate is valid
+ * - 0th certificate is a valid code signing certificate
+ */
+class EXUpdatesCertificateChain {
+  // ASN.1 path to the extended key usage info within a CERT
+  static let EXUpdatesCodeSigningCertificateExtendedUsageCodeSigningOID = "1.3.6.1.5.5.7.3.3"
+  
+  private var certificateStrings: [String]
+  
+  public required init(certificateStrings: [String]) throws {
+    self.certificateStrings = certificateStrings
+  }
+  
+  public func codeSigningCertificate() throws -> Certificate {
+    if (certificateStrings.isEmpty) {
+      throw EXUpdatesCodeSigningError.CertificateEmptyError
+    }
+    
+    let certificateChain = try certificateStrings.map { certificateString throws in
+      try EXUpdatesCertificateChain.constructCertificate(certificateString: certificateString)
+    }
+    try certificateChain.validateChain()
+    
+    let leafCertificate = certificateChain.first!
+    let (_, x509lLeafCertificate) = leafCertificate
+    if (!x509lLeafCertificate.isCodeSigningCertificate()) {
+      throw EXUpdatesCodeSigningError.CertificateMissingCodeSigningError
+    }
+    
+    return leafCertificate
+  }
+  
+  private static func constructCertificate(certificateString: String) throws -> Certificate {
+    guard let certificateData = certificateString.data(using: .utf8) else { throw EXUpdatesCodeSigningError.CertificateParseError }
+    
+    guard let certificateDataDer = decodeToDER(pem: certificateData) else {
+      throw EXUpdatesCodeSigningError.CertificateParseError
+    }
+    
+    let x509Certificate = try X509Certificate(der: certificateDataDer)
+    
+    guard x509Certificate.checkValidity() else {
+      throw EXUpdatesCodeSigningError.CertificateValidityError
+    }
+    
+    guard let secCertificate = SecCertificateCreateWithData(nil, certificateDataDer as CFData) else {
+      throw EXUpdatesCodeSigningError.CertificateParseError
+    }
+    
+    return (secCertificate, x509Certificate)
+  }
+  
+  private static let beginPemBlock = "-----BEGIN CERTIFICATE-----"
+  private static let endPemBlock   = "-----END CERTIFICATE-----"
+  
+  /**
+   * Mostly from ASN1Decoder with the fix for disallowing multiple certificatess in the PEM.
+   */
+  private static func decodeToDER(pem pemData: Data) -> Data? {
```

Comment:
According to https://datatracker.ietf.org/doc/html/rfc7468#section-2, whitespace is supposed to be ignored in PEM files. For certificates we control, this isn't a problem, but it's apparently valid to have trailing whitespace at the end of lines or no newlines at all.

Looking for the BEGIN and END blocks and stripping out whitespace in between seems about as complex as the code here, worth considering.

Reply:
Definitely a possibility. I think I prefer to keep it as-is for now since this is what we rely upon in ASN1Decoder anyways, so it's easier to have only one implementation.